### PR TITLE
FUSETOOLS-2531 - provide a server adapter for Fuse 7 

### PR DESCRIPTION
**Pull request by lhein**
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

### CODE REVIEW

**Comment by lhein** [comment]
just figured that this PR needs #1108 merged to actually be able to build.

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		return DEFAULT_FUSEESB_SSH_PASSWORD;
+	}
+	
+	@Override
+	public ILaunchConfigConfigurator getLaunchConfigurator() throws CoreException {
+		return new Fuse7xStartupLaunchConfigurator(getServer());
+	}
+	
+	@Override
+	public IStatus validate() {
```
[the remark this answers is not in the document]
Comment (lhein) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java]: We just override the validation done by the super class. Fuse6 delegate does not need to do that.

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse7xStartupLaunchConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.eclipse.wst.server.core.IServer;
+import org.fusesource.ide.server.fuse.core.Activator;
+import org.fusesource.ide.server.fuse.core.util.FuseToolingConstants;
+import org.fusesource.ide.server.karaf.core.runtime.IKarafRuntime;
+import org.fusesource.ide.server.karaf.core.server.subsystems.Karaf4xStartupLaunchConfigurator;
+
+/**
+ * @author lhein
+ */
+public class Fuse7xStartupLaunchConfigurator extends
```
[the remark this answers is not in the document]
Comment (lhein) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse7xStartupLaunchConfigurator.java]: both classes have different super classes. Is it really worth a restructuring?

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/runtime/FuseESBRuntimeDelegate.java [hunk 3]
```diff
@@ -12,41 +12,43 @@
...
+		if (!runtimeId.toLowerCase().endsWith(String.format("fuseesb.runtime.%s", versionStartString)) && !isNewerUnsupportedVersion(runtimeVersion)) {
+			return new Status(Status.ERROR, Activator.PLUGIN_ID, "Runtime type " + runtimeId + " is not compatible with found version " + runtimeVersion);
+		}
 		return Status.OK_STATUS;
 	}
+	
+	@Override
+	public IExecutionEnvironment getMinimumExecutionEnvironment() {
+		if (getRuntime().getRuntimeType().getVersion().startsWith(FuseToolingConstants.FUSE_VERSION_7X)) {
+			return EnvironmentsManager.getDefault().getEnvironment("JavaSE-1.8"); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (lhein) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/runtime/FuseESBRuntimeDelegate.java]: Where is that class located? My class search doesn't show it.

[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/runtime/FuseESBRuntimeDelegate.java]: org.eclipse.jdt.launching.JavaRuntime.getExecutionEnvironmentsManager().getEnvironment(String)

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse7xStartupLaunchConfigurator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse7xStartupLaunchConfigurator.java]: deduplicating more than 100 lines of code? I tend to say yes

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java]: ok

**Comment by apupier** [comment]
testPR

**Comment by lhein** [comment]
This will not build until #1108 has been merged into master and this PR rebased.

**Comment by lhein** [comment]
@apupier can you give it another try to see if your error from the screenshots above is gone or not?

**Comment by apupier** [comment]
last build has failed with 2 test in errors:
image

one is a known unstable test, the other one seems related to this PR https://fusesource-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/jbosstools-fuse_PullRequest/228/testReport/junit/org.fusesource.ide.server.tests.bean/FuseESBServerBean7xIT/testFuseESB_0_/

Error Message

expected:<7.0.0.[fuse-]000080> but was:<7.0.0.[redhat-0]000080>

Stacktrace

junit.framework.ComparisonFailure: expected:<7.0.0.[fuse-]000080> but was:<7.0.0.[redhat-0]000080>
	at junit.framework.Assert.assertEquals(Assert.java:100)
	at junit.framework.Assert.assertEquals(Assert.java:107)
	at junit.framework.TestCase.assertEquals(TestCase.java:269)
	at org.fusesource.ide.server.tests.bean.FuseESBServerBean7xIT.testFuseESB(FuseESBServerBean7xIT.java:72)

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java [hunk 4]
```diff
@@ -24,10 +30,19 @@
...
+	public static final String FUSEESB_62 = "org.fusesource.ide.fuseesb.runtime.62";
+	public static final String FUSEESB_63 = "org.fusesource.ide.fuseesb.runtime.63";
+	public static final String FUSEESB_6X = "org.fusesource.ide.fuseesb.runtime.6x";
+	
+	public static final String FUSEESB_70 = "org.fusesource.ide.fuseesb.runtime.70";
+	public static final String FUSEESB_7X = "org.fusesource.ide.fuseesb.runtime.7x";
 	
 	public static final String[] SUPPORTED_6X_RUNTIMES = new String[] {
-		FUSEESB_60, FUSEESB_61
+		FUSEESB_60, FUSEESB_61, FUSEESB_62, FUSEESB_63
```
Comment (apupier) [inline servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: it is often more convenient to use List<String>
	public static final List<String> SUPPORTED_6X_RUNTIMES = Arrays.asList(FUSEESB_60, FUSEESB_61, FUSEESB_62, FUSEESB_63);

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java [hunk 5]
```diff
@@ -24,10 +30,19 @@
...
+	};
+	
+	public static final String[] SUPPORTED_7X_RUNTIMES = new String[] {
```
Comment (apupier) [inline servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: it is often more convenient to use List<String>
public static final List<String> SUPPORTED_7X_RUNTIMES =  Arrays.asList(FUSEESB_70);

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java [hunk 6]
```diff
@@ -52,13 +67,42 @@ public static boolean create6xRuntimeMock(String runtimeId, IPath runtimePath) {
...
+	
 	private static boolean isSupported6xRuntime (String runtimeId) {
 		for (String id : SUPPORTED_6X_RUNTIMES) {
 			if (id.equals(runtimeId)) return true;
 		}
 		return false;
 	}
 	
+	private static boolean isSupported7xRuntime (String runtimeId) {
+		for (String id : SUPPORTED_7X_RUNTIMES) {
```
Comment (apupier) [inline servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: if using List<String>, it can be written simpler:
return SUPPORTED_7X_RUNTIMES.contains(runtimeId);

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java [hunk 7]
```diff
@@ -52,13 +67,42 @@ public static boolean create6xRuntimeMock(String runtimeId, IPath runtimePath) {
...
+			} catch (CoreException ex) {
+				runtimeCreated = false;
+			}
+		}		
+		
+		return runtimeCreated;
+	}
+	
 	private static boolean isSupported6xRuntime (String runtimeId) {
 		for (String id : SUPPORTED_6X_RUNTIMES) {
```
Comment (apupier) [inline servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: if using List<String>, it can be written simpler:
return SUPPORTED_6X_RUNTIMES.contains(runtimeId);

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/bean/FuseESBServerBean7xIT.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	public FuseESBServerBean7xIT(String runtimeType) {
+		this.fRuntimeType = runtimeType;
+	}
+	
+	/**
+	 * returns the runtime types to test
+	 * @return
+	 */
+	@Parameters
+	public static Collection<Object[]> data() {
```
Comment (apupier) [inline servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/bean/FuseESBServerBean7xIT.java]: if using List instead of Array, you can write it
	@Parameters
	public static Iterable<? extends Object> data() {
		return FuseESBMockRuntimeCreationUtil.SUPPORTED_7X_RUNTIMES;
	}

**Comment by lhein** [comment]
> I cannot connect to the shell (but I think that it is due to a bug in Fuse right?)

Correct. The user lacks a role. That is not a problem of the tooling and has been already adressed by sustaining team.

**Comment by lhein** [comment]
> I get this log when starting the server from UI:
2017-09-28 18:14:08,056 features-1-thread-1 WARN Unable to instantiate org.fusesource.jansi.WindowsAnsiOutputStream
2017-09-28 18:14:08,059 features-1-thread-1 WARN Unable to instantiate org.fusesource.jansi.WindowsAnsiOutputStream
2017-09-28 18:14:08,125 CM Configuration Updater (ManagedService Update: pid=[org.ops4j.pax.logging]) WARN Unable to instantiate org.fusesource.jansi.WindowsAnsiOutputStream

I fear I can't help with that. On Linux it works fine. Maybe you can check whats the issue?

**Comment by lhein** [comment]
> i'm wondering if we should not inheritance between Karaf 2.x, 3.x and 4.x after a quick eye on them it seems that they are overriding the same method every time and don't reuse what is just above. Did I miss something?

Honestly I can't follow what you state here. 3x dropped derby and hence has different vmarguments. 4x extends 3x as its also not using derby but in addtion has a different findJars method. Why should they not extend as they do? I don't get the point.

**Comment by lhein** [comment]
> shouldn't we have the Fuse6xStartupLaunchConfigurator extending Karaf2xStartupLaunchConfigurator?

There is not really an advantage to inherit from 2x. Why should that be better than inheriting from the common base class?

**Comment by apupier** [comment]
testPR

**Comment by lhein** [comment]
@apupier on Oct 4th we got the feature freeze. Even if you still find issues in this PR I opt for merging it and open another Jira for the fixes otherwise we will not have Fuse 7 support in the 10.1 release.

**Comment by apupier** [comment]
/!\ on going discussion for not including it in this release (due to date of publicly available Fuse 7 on Karaf)

**Comment by apupier** [comment]
- rebased with master as there was a conflict
- provided more tests (which helped to detect...)
- some issues with doOverrides

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java [hunk 9]
```diff
@@ -14,49 +14,31 @@
...
-		vmArguments.append(SPACE + "-Dkaraf.data=" + QUOTE + karafInstallDir + SEPARATOR + "data" + QUOTE);
-		vmArguments.append(SPACE + "-Dkaraf.etc=" + QUOTE + karafInstallDir + SEPARATOR + "etc" + QUOTE);
-		vmArguments.append(SPACE + "-Djava.io.tmpdir=" + QUOTE + karafInstallDir + SEPARATOR + "data" + SEPARATOR + "tmp" + QUOTE);
-		vmArguments.append(SPACE + "-Djava.util.logging.config.file=" + QUOTE + karafInstallDir + SEPARATOR + "etc" + SEPARATOR + "java.util.logging.properties" + QUOTE);
-		
-		return vmArguments.toString();
+	public String getVMArguments(String karafInstallDir, String endorsedDirs, String extDirs) {
+		StringBuilder sb = new StringBuilder(super.getVMArguments(karafInstallDir, endorsedDirs, extDirs));
+		sb.append(SPACE + "-XX:PermSize=16M -XX:MaxPermSize=128M");
+		return sb.toString();
```
Comment (lhein) [inline servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java]: I don't think that this is easier to read than what I had here before.

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java [hunk 10]
```diff
@@ -14,49 +14,31 @@
...
 import org.fusesource.ide.foundation.core.util.Strings;
 import org.fusesource.ide.server.fuse.core.util.FuseToolingConstants;
-import org.fusesource.ide.server.karaf.core.server.subsystems.BaseKarafStartupLaunchConfigurator;
+import org.fusesource.ide.server.karaf.core.server.subsystems.Karaf2xStartupLaunchConfigurator;
 
 /**
  * @author lhein
  */
-public class Fuse6xStartupLaunchConfigurator extends BaseKarafStartupLaunchConfigurator {
+public class Fuse6xStartupLaunchConfigurator extends Karaf2xStartupLaunchConfigurator {
```
Comment (lhein) [inline servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java]: There is no advantage to inherit from Karaf2x class. Why did you change it?

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import org.fusesource.ide.server.karaf.core.server.Karaf4xServerDelegate;
+import org.jboss.ide.eclipse.as.core.server.ILaunchConfigConfigurator;
+
+/**
+ * @author lhein
+ */
+public class Fuse7xServerDelegate extends Karaf4xServerDelegate {
+	
+	public static final String DEFAULT_FUSEESB_SSH_USER = "admin";
+	public static final String DEFAULT_FUSEESB_SSH_PASSWORD = "admin";
```
[the remark this answers is not in the document]
Comment (lhein) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java]: should we mark as false positive? I think that can be safely ignored.

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse6xServerDelegate.java [hunk 12]
```diff
@@ -18,7 +18,7 @@
 /**
  * @author lhein
  */
-public class FuseESBServerDelegate extends KarafServerDelegate {
+public class Fuse6xServerDelegate extends KarafServerDelegate {
 	
 	public static final String DEFAULT_FUSEESB_SSH_USER = "admin";
 	public static final String DEFAULT_FUSEESB_SSH_PASSWORD = "admin";
```
[the remark this answers is not in the document]
Comment (lhein) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse6xServerDelegate.java]: should we mark as false positive? I think that can be safely ignored.

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java]: yes i think that we can mark it as ignore.
it might worth factorizing the credentials for Fuse Fuse6x and Fuse7x

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse6xServerDelegate.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse6xServerDelegate.java]: yes i think that we can mark it as ignore.
it might worth factorizing the credentials for Fuse Fuse6x and Fuse7x

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java (hunk 9 above)
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java]: it allows to see which parameters are added for/since a specific version

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java (hunk 10 above)
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/subsystems/Fuse6xStartupLaunchConfigurator.java]: I think that it better reflect the reality (Fuse 6 is based on Karaf 2.x).

**Comment by lhein** [comment]
Rebased

**Comment by apupier** [comment]
if I remember well, it looks alike 2 commits with there modifications were lost during the rebase.

I recreated the branch from my local environment here: https://github.com/apupier/jbosstools-fuse/tree/lhein-FUSETOOLS-2531
The 2 commits are these ones:
- https://github.com/apupier/jbosstools-fuse/commit/3e295448bca65631042104c6dc6e45a07b513ac7
- https://github.com/apupier/jbosstools-fuse/commit/29c3c0dbb35bc92278cc965510cbaba9843b3f44

**Comment by lhein** [comment]
I have no idea how those 2 commits were abandoned. :(

**Comment by lhein** [comment]
Ready for Review

File: servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/AbstractFuseESBRuntimeComposite.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+import org.fusesource.ide.server.karaf.ui.runtime.KarafWizardDataModel;
+import org.fusesource.ide.server.karaf.ui.runtime.v2x.KarafRuntimeComposite2x;
+
+/**
+ * @author lheinema
+ *
+ */
+public abstract class AbstractFuseESBRuntimeComposite extends KarafRuntimeComposite2x {
+	/**
+	 * constructor
```
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/AbstractFuseESBRuntimeComposite.java]: What do you think?
Worth removing it? or to provide a complete one with more details?

[the remark this answers is not in the document]
Comment (lhein) [reply servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/AbstractFuseESBRuntimeComposite.java]: will remove it

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java (hunk 6 above)
Comment (apupier) [reply servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: up

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java (hunk 7 above)
Comment (apupier) [reply servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: up

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java (hunk 5 above)
Comment (apupier) [reply servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: up

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java (hunk 4 above)
Comment (apupier) [reply servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: up

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/bean/FuseESBServerBean7xIT.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2013 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (apupier) [reply servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/bean/FuseESBServerBean7xIT.java]: up

**Comment by apupier** [comment]
there was discussion to not display the Fuse 7 Server adapter as there was no Fuse 7 Karaf-based runtime publicly available: https://github.com/jbosstools/jbosstools-fuse/pull/1107#issuecomment-333462306
Do we have a Fuse 7 Tech Preview Karaf-based publicly available now? or do we know the date fo release? Will it be before next Fuse Tooling Milestone release? Final release?

**Comment by lhein** [comment]
The releases will happen around the same time as far as I know.

**Comment by apupier** [comment]
> The releases will happen around the same time as far as I know.

the same time of the 10.2.0AM1 or of the 10.2.0.Final?

**Comment by lhein** [comment]
I am talking about final releases. If we want to support early versions then we could do that by providing appropriate stacks.yaml changes pointing to those pre-release runtimes. Personally I'd prefer to only put finals into that file.

**Comment by apupier** [comment]
> I am talking about final releases. If we want to support early versions then we could do that by providing appropriate stacks.yaml changes pointing to those pre-release runtimes. Personally I'd prefer to only put finals into that file.

the problem which triggered discussion to hide the Fuse 7 Server adapter was about no possibility to have a FUSE 7 runtime for external users of Red Hat,. It was not about having it integrated in the UI for easy download.

And, if we have a tech preview, I think that it would be worthy to provide it in the download list.

So seems that:
- need to check the release date of Fuse 7 based on Karaf
- do we want to have Fuse 7 Server Adapter visible in the Milestone releases of Fuse Tooling although no public build of Fuse 7 runtime Karaf-based are available?

**Comment by lhein** [comment]
Afaik Fuse Tooling / JBT milestone releases are not made public. Please correct me if I am wrong. So there is no point in hiding that for internal releases.

**Comment by apupier** [comment]
> Afaik Fuse Tooling / JBT milestone releases are not made public. Please correct me if I am wrong. So there is no point in hiding that for internal releases.

All Fuse tooling/JBoss Tools milestones are public, they are announced on the Jboss Tools blog, for instance: https://tools.jboss.org/blog/4.5.1.am2-for-oxygen.1.html

**Comment by lhein** [comment]
Ok, so that mail is misleading then. So I see only 2 options to satisfy your requirements:

1. hide the entry by for now by commenting appropriate plugin.xml section
2. provide download links for selected versions of Fuse 7 runtime

IMO we are making a much to big deal out of this. Fuse 7 will show up in the list, thats true. BUT: it won't be downloadable from within the IDE. If a user somehow grabs a Fuse 7 pre-release / Milestone and extracts it to his computer he will be able to use our Fuse 7 adapter with that runtime. 

I strongly opt for not doing anything about it. Its absolutely valid.

File: servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/AbstractFuseESBRuntimeComposite.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+import org.fusesource.ide.foundation.core.util.Strings;
+import org.fusesource.ide.server.fuse.ui.Messages;
+import org.fusesource.ide.server.karaf.ui.runtime.KarafWizardDataModel;
+import org.fusesource.ide.server.karaf.ui.runtime.v2x.KarafRuntimeComposite2x;
+
+/**
+ * @author lheinema
+ *
+ */
+public abstract class AbstractFuseESBRuntimeComposite extends KarafRuntimeComposite2x {
```
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/AbstractFuseESBRuntimeComposite.java]: ok to leave it

File: servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/FuseESBRuntimeComposite7x.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.wst.server.ui.wizard.IWizardHandle;
+import org.fusesource.ide.server.fuse.ui.Messages;
+import org.fusesource.ide.server.karaf.ui.runtime.KarafWizardDataModel;
+
+
+/**
+ * @author lhein
+ */
+public class FuseESBRuntimeComposite7x extends AbstractFuseESBRuntimeComposite {
```
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.ui/src/org/fusesource/ide/server/fuse/ui/runtime/fuseesb/FuseESBRuntimeComposite7x.java]: ok to leave it

File: servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java [hunk 17]
```diff
@@ -24,11 +32,15 @@
...
+	public static final String FUSEESB_6X = "org.fusesource.ide.fuseesb.runtime.6x";
 	
-	public static final String[] SUPPORTED_6X_RUNTIMES = new String[] {
-		FUSEESB_60, FUSEESB_61
-	};
+	public static final String FUSEESB_70 = "org.fusesource.ide.fuseesb.runtime.70";
+	public static final String FUSEESB_7X = "org.fusesource.ide.fuseesb.runtime.7x";
+	
+	public static final List SUPPORTED_6X_RUNTIMES = Arrays.asList(FUSEESB_60, FUSEESB_61, FUSEESB_62, FUSEESB_63);
+	public static final List SUPPORTED_7X_RUNTIMES = Arrays.asList(FUSEESB_70);
```
[the remark this answers is not in the document]
Comment (apupier) [reply servers/tests/org.fusesource.ide.server.tests/src/main/java/org/fusesource/ide/server/tests/util/FuseESBMockRuntimeCreationUtil.java]: see https://github.com/lhein/fuseide/pull/46

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse7xServerDelegate.java]: see https://github.com/lhein/fuseide/pull/47 for a proposal of factorization

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse6xServerDelegate.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/Fuse6xServerDelegate.java]: see https://github.com/lhein/fuseide/pull/47 for a proposal of factorization

File: servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/FuseServerConstants.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Contributors:
+ *     Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package org.fusesource.ide.server.fuse.core.server;
+
+public class FuseServerConstants {
+	
+	public static final String DEFAULT_FUSEESB_SSH_USER = "admin";
+	public static final String DEFAULT_FUSEESB_SSH_PASSWORD = "admin";
```
[the remark this answers is not in the document]
Comment (apupier) [reply servers/plugins/org.fusesource.ide.server.fuse.core/src/org/fusesource/ide/server/fuse/core/server/FuseServerConstants.java]: will be marked as false-positive on Sonar instance

**Comment by apupier** [comment]
take care to wait for code unfreeze (planned Tuesday next week)
